system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -3168,6 +3325,95 @@ class CLdapClient : public CInterface, implements ILdapClient
+        const char* prefix, const char* resourceName, unsigned extraNameFilter)
+    {
+        IArrayOf<ISecResource> resources;
+        getResourcesEx(rtype, basedn, prefix, resourceName, resources);
+
+        Owned<IPTree> resourceTree = createPTree("Resources");
+        ForEachItemIn(i, resources)
+        {
+            ISecResource& resource = resources.item(i);
+            const char* resourceName = resource.getName();
+            if (!resourceName || !*resourceName)
+                continue;
+            if (checkResourceNameExtraFilter(rtype, resourceName, extraNameFilter))
+                addResourceTree(resourceName, resource.getDescription(), resourceTree);
+        }
+        return resourceTree->getElements("*");
+    }
+
+    bool checkResourceNameExtraFilter(SecResourceType rtype, const char* name, unsigned extraNameFilter)
+    {
+        if((rtype == RT_FILE_SCOPE) && (extraNameFilter & RF_RT_FILE_SCOPE_FILE) && strieq(name, "file"))
+            return false;
+        if((rtype == RT_MODULE) && (extraNameFilter & RF_RT_MODULE_NO_REPOSITORY) && strnicmp(name, "repository.", 11))
+            return false;
+        return true;
+    }
+
+    void addResourceTree(const char* name, const char* desc, IPropertyTree* elements)
+    {
+        if (!name || !*name)
+            return;
+
+        Owned<IPTree> element = createPTree();
+        element->addProp(getResourceFieldNames(RFName), name);
+        if (desc && *desc)
+            element->addProp(getResourceFieldNames(RFDesc), desc);
+        elements->addPropTree("Resource", element.getClear());
+    }
+
+    ISecItemIterator* getResourcesSorted(SecResourceType rtype, const char * basedn, const char* resourceName, unsigned extraNameFilter,
+        ResourceField* sortOrder, const unsigned pageStartFrom, const unsigned pageSize, unsigned* total, __int64* cacheHint)
+    {
+        class CElementsPager : public CSimpleInterface, implements IElementsPager
+        {
+            ILdapClient* ldapClient;
+            StringAttr sortOrder, basedn, resourceName;
+            SecResourceType rtype;
+            unsigned extraNameFilter;
+
+        public:
+            IMPLEMENT_IINTERFACE_USING(CSimpleInterface);
+
+            CElementsPager(ILdapClient* _ldapClient, SecResourceType _rtype, const char * _basedn, const char* _resourceName,
+                unsigned _extraNameFilter, const char*_sortOrder) : ldapClient(_ldapClient), rtype(_rtype), basedn(_basedn),
+                resourceName(_resourceName), extraNameFilter(_extraNameFilter), sortOrder(_sortOrder) { };
+            virtual IRemoteConnection* getElements(IArrayOf<IPropertyTree>& elements)
+            {
+                StringArray unknownAttributes;
+                Owned<IPropertyTreeIterator> iter = ldapClient->getResourceIterator(rtype, basedn.get(), "", resourceName.get(), extraNameFilter);
+                sortElements(iter, sortOrder.get(), NULL, NULL, unknownAttributes, elements);
```

Comment:
I don't really understand this code but is this always sorting locally having paged through all the values retrieved form ldap? If so doesn't it defeat the purpose of paging the info from ldap?